From 1ca472c7208a7f72ab8a61aa7fe5fe1954fc345b Mon Sep 17 00:00:00 2001
From: David Vrabel <david.vrabel@citrix.com>
Date: Tue, 3 May 2016 17:15:38 +0100
Subject: [PATCH] x86: show correct code in CPU state

When showing the CPU state (e.g., after a crash) the dump of code
around RIP is incorrect.

Incorrect:

    Xen code around <ffff82d0801113cf> (...):
     00 c6 c1 ee 08 48 c1 e0 <04> 03 04 f1 8b ...
     ^^ Uninitialized         ^^ Missing 0x48

Correct:

    Xen code around <ffff82d0801113cf> (...):
     c6 c1 ee 08 48 c1 e0 04 <48> 03 04 f1 8b ...

When coping the bytes before RIP, the destination was off-by-one.

Signed-off-by: David Vrabel <david.vrabel@citrix.com>
Reviewed-by: Andrew Cooper <andrew.cooper3@citrix.com>
Release-acked-by: Wei Liu <wei.liu2@citrix.com>
diff --git a/xen/arch/x86/traps.c b/xen/arch/x86/traps.c
index 6ab85d8..db4180d 100644
--- a/xen/arch/x86/traps.c
+++ b/xen/arch/x86/traps.c
@@ -148,7 +148,7 @@ static void show_code(const struct cpu_user_regs *regs)
                   : "=&c" (missing_before),
                     "=&D" (tmp), "=&S" (tmp)
                   : "0" (ARRAY_SIZE(insns_before)),
-                    "1" (insns_before + ARRAY_SIZE(insns_before)),
+                    "1" (insns_before + ARRAY_SIZE(insns_before) - 1),
                     "2" (regs->rip - 1));
     clac();
 
